### DISCUSSION TITLE
Сборка руководств

### DISCUSSION THREAD

**Discussion by eugenekogan-spb** [fhir-ru/core#264] [General]
Пока программисты отлаживают сборку руководств на Fhir.ru, надо наладить альтернативную сборку.
Для этого выложим fsh-тексты в репозиторий и наладим работу паблишера.

Выяснилось, что в данный момент сборщик на Fhir.ru собирает fsh-тексты, но не сохраняет их в репозитории. Ждём ответа разработчиков.

@Vrodvi

**Comment by eugenekogan-spb** [comment]
От О.Пензина: виртуальная машина с публикатором и инструментами fsh готова, выложена здесь: https://drive.google.com/file/d/1Xsgk488qB-xASO3zxpUAtjyxWXeFz6KX/view?usp=sharing
Будут ли какие-то хотя бы краткие инструкции по использованию?
@OlegPenzin

**Comment by Vrodvi** [comment]
@eugenekogan-spb 

По результатам договорённости с совещания https://fhir.ru/meetings.2024-06-04 осуществил проверку возможности ведения проекта руководства 
в репозитории (https://github.com/fhir-ru/ru.core)
, где уже имеются папки проектов 
ру-кор (https://github.com/fhir-ru/ru.core/tree/main/RuCoreIG) 
и ру-лаб (https://github.com/fhir-ru/ru.core/tree/main/RuLabIG).

Для этого в образе виртуальной машины:

1) клонировал указанный репозиторий (https://github.com/fhir-ru/ru.core)
2) в папке проекта ру-кор добавил:
- файл профиля организации (в нём сам профиль и пример, соответствующий этому профилю) - RuCoreIG/intput/fsh/CoreOrganization.fsh
- файл заголовка|intro (оформленного по описанию https://github.com/fhir-ru/core/discussions/235#discussioncomment-7764147) для профиля организации с фрагментом описания из https://fhir.ru/core.profile.Organization - RuCoreIG/intput/pagecontent/StructureDefinition-core-organization-intro.md
- файл подвала|notes (оформленного по описанию https://github.com/fhir-ru/core/discussions/235#discussioncomment-7764147) для профиля организации с текстом FSH это профиля в качестве наполнения - RuCoreIG/intput/pagecontent/StructureDefinition-core-organization-notes.md
![image](https://github.com/fhir-ru/core/assets/92928137/e9bfb68c-ddf0-4073-8f7c-24ecb8c808ef)
3) запущен инструмент генерации FHIR-профилей SUSHI - успешно.
![image](https://github.com/fhir-ru/core/assets/92928137/60108462-92aa-4242-a2dd-287b75212b0b)
4) запущен инструмент генерации сайта руководства (publisher.jar, _genonce.sh) - успешно.
При генерации страницы профиля организации успешно добавлены описания заголовка и подвала:
![image](https://github.com/fhir-ru/core/assets/92928137/fab8bed1-517b-4935-a1ad-138828ba0be4)
5) Осуществлена попытка зафиксировать внесённые изменения обратно в репозиторий - частично успешно, т.к. процесс фиксации изменений запускается, но поскольку в данном репозитории у меня нет прав делать прямые фиксации, то процесс прерывается:
![2024-06-04_15h42_53](https://github.com/fhir-ru/core/assets/92928137/38ab16b0-8cab-40f3-be5a-6fcce07410e4)
![image](https://github.com/fhir-ru/core/assets/92928137/9294f1ec-ec7a-4c25-a3af-1fb484f39738)

Таким образом продемонстрирована возможность осуществлять редактирование проекта руоководства в многопользовательском режиме с использованием подготовленного образа виртуальной машины с настроенным програмным обеспечением HL7 (SUSHI, Publisher) и настроенным общим репозиторием проекта в GitHub.

Обновлённый образ виртуальной машины подготовлен, но разместить его у меня нет возможности из-за большого объёма (7,21 ГБ), образ будет передан @OlegPenzin для последующей публикации (вероятно по тому же адресу https://drive.google.com/file/d/1Xsgk488qB-xASO3zxpUAtjyxWXeFz6KX/view?usp=sharing).

**Reply by Vrodvi** [reply]
Обновлённый образ виртуальной машины опубликован по новому адресу:
https://drive.google.com/file/d/1N6h5afl-JXGdAw9mGV5LxlV0Ew1pq2u3/view?usp=sharing

**Comment by eugenekogan-spb** [comment]
@Vrodvi я залил на ру-кор (https://github.com/fhir-ru/ru.core/tree/main/RuCoreIG) файлы с fsh-текстами артефактов (папка fsh) и файлы с хедерами для профилей (папка pagecontent).
Проверьте пожалуйста как собирается руководство.

**Reply by Vrodvi** [reply]
Запустил в виртуальной машине обновление репозитория - ваши изменения успешно загрузились.
Запустил для начала просто преобразование FSH в FHIR через инструмент SUSHI - обнаружено 24 ошибки.
Собрать сайт руководства при наличии ошибок не получится, Publisher даже не запустится пока имеются ошибки SUSHI.

Ошибки в логе выглядят следующим образом:
```
error No element found at path payor for CaretValueRule in Core_Coverage, skipping rule
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreCoverage.fsh
  Line: 8 - 9
error No element found at path payor for CardRule in Core_Coverage, skipping rule
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreCoverage.fsh
  Line: 10
error No element found at path payor for OnlyRule in Core_Coverage, skipping rule
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreCoverage.fsh
  Line: 11
error No element found at path address for CardRule in Core_Organization, skipping rule
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 14 - 15
error No element found at path address for CaretValueRule in Core_Organization, skipping rule
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 16
error No element found at path address for OnlyRule in Core_Organization, skipping rule
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 17
error The element or path you referenced does not exist: payor
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreCoverage.fsh
  Line: 60 - 61
error Element Coverage.kind has minimum cardinality 1 but occurs 0 time(s).
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreCoverage.fsh
  Line: 55 - 70
error The element or path you referenced does not exist: address
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 57
error The element or path you referenced does not exist: address.use
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 58
error The element or path you referenced does not exist: address.type
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 59
error The element or path you referenced does not exist: address.city
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 60
error The element or path you referenced does not exist: address.text
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 61
error The element or path you referenced does not exist: address.postalCode
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 62
error The element or path you referenced does not exist: address.country
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 63
error The element or path you referenced does not exist: address.extension[fias]
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 64
error The element or path you referenced does not exist: address.extension[fias].extension[aoguid]
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 65
error The element or path you referenced does not exist: address.extension[fias].extension[aoguid].valueIdentifier
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 66
error The element or path you referenced does not exist: address.extension[fias].extension[aoguid].valueIdentifier.value
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 67
error The element or path you referenced does not exist: address.extension[fias].extension[aoguid].valueIdentifier.system
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 68
error The element or path you referenced does not exist: address.extension[fias].extension[houseguid]
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 69
error The element or path you referenced does not exist: address.extension[fias].extension[houseguid].valueIdentifier
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 70
error The element or path you referenced does not exist: address.extension[fias].extension[houseguid].valueIdentifier.value
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 71
error The element or path you referenced does not exist: address.extension[fias].extension[houseguid].valueIdentifier.system
  File: /home/fhir/Documents/ru.core/RuCoreIG/input/fsh/CoreOrganization.fsh
  Line: 72
```

![image](https://github.com/fhir-ru/core/assets/92928137/61adbf60-0855-4ce6-be8e-25444d4922e1)

**Reply by eugenekogan-spb** [reply]
1) CoreCoverage.payor - опечатка, исправлено.
2) CoreCoverage.kind - этого атрибута нет в R4. Видимо, sushi предполагает, что это тексты R5. 
3) Organization.address - этот атрибут есть в R4, но нет в R5. Видимо, та же проблема.

**Reply by Vrodvi** [reply]
Кругом загадки. 
Хотя в перечне руководств и написано R4, но фактически сайт руководства сейчас собирается под R5, потому что именно так настроено сейчас в репозитории.
![image](https://github.com/fhir-ru/core/assets/92928137/15e64779-35f5-4d34-99aa-137f9a0a9efd)

**Reply by Vrodvi** [reply]
Если настройку руководства переключить на версию 4.3.0 (а также если добавить один недостающий профиль Core_Organization_Payer), то ошибки уходят и руководство успешно формируется.

![image](https://github.com/fhir-ru/core/assets/92928137/aeb4e4a8-e4b6-4b9f-9b5d-3b9acb3aa141)

**Reply by eugenekogan-spb** [reply]
Добавил недостающий профиль, поправил настройку руководства.

**Comment by eugenekogan-spb** [comment]
После загрузки файлов мне стали приходить сообщения о каких-то ошибках докера:
<img width="1416" alt="image" src="https://github.com/fhir-ru/core/assets/62525329/4108eefd-3b7c-4763-a3d6-894166fdeb4c">

**Reply by Vrodvi** [reply]
Возможно это проблемы с автоматизированным процессом сборки сайта, который настроили хелссамураи. Мы обсуждали, что ручные правки в папке руководства могут нарушить работу этого процесса. Тут нужна консультация авторов этой автоматизации.

**Comment by eugenekogan-spb** [comment]
10 июня В.Родионов провел сборку руководства RuCore, на основании fsh-файлов и конфигураций из папки https://github.com/fhir-ru/ru.core/tree/main/RuCoreIG. QA report выложен [здесь](https://docs.google.com/document/d/1MdFMnVl8Axhfqdr7FrH1T5ex29fjd9n0/edit?usp=sharing&ouid=107253868441487824494&rtpof=true&sd=true). Он содержит 49 ошибок и 70 предупреждений.
Fsh-тексты изначально проходили проверку fsh online, поэтому есть надежда, что есть какие-то ошибки сборки, дающие такие множественные последствия.
Может ли кто-то помочь их разыскать?
@Vrodvi

**Reply by Vrodvi** [reply]
Большая часть ошибок - не связаны со сборкой, а либо просто опечатки в наполнении либо некорректное профилирование:

1) пример опечаток - в примере Coverage указаны несуществующие идентификаторы:
```
* payor = Reference(core-patient-payer-instance-ful)
* beneficiary = Reference(core-patient-instance-ful)
```
- в идентификаторе пропущена одна буква "l', должно быть ```-full```

2) пример некорректного профилирования - в организации атрибут связан со справочником напрямую, а может быть связан только с набором значений:
```
  * type = #TAX
  * type from http://terminology.hl7.org/CodeSystem/v2-0203
```
что удивительно - прямо по соседству запрофилировано правильно:
```
  * type from $Ядро_НаборЗначений_ДокументыУдостоверяющиеЛичность (required)
  * type = $МЗРФ_Справочник_ДокументыУдостоверяющиеЛичность#1
```

![image](https://github.com/fhir-ru/core/assets/92928137/67fb4e44-ce58-4d94-b8f2-b062de078fa6)

**Reply by Vrodvi** [reply]
@eugenekogan-spb 
Почти все ошибки были из-за кривого профилирования (ещё было два ValueSet, которые недоступны и для базовой спецификации, и они просочились в IG - если их переопределить, то и эти ошибки уходят).
Поправил грубо - чтобы показать, что проблема не в сборке.
На многочисленные замечания (warning) пока не смотрел - там чисто стилистическое.
Исправленный вариант Input и Output прикладываю:
[input_output_для_RuCoreIG.zip](https://github.com/user-attachments/files/15974742/input_output_._RuCoreIG.zip)

**Reply by eugenekogan-spb** [reply]
Я залил файлы из input в репозиторий, дополнительно поменял title так чтобы были видны и русские, и латинские названия - это решение было принято совместно с @OlegPenzin на встрече 02.07.24.
Замечена странность: в файле CoreOrganizationPayer.fsh есть только заголовки (Title, description), но нет самих ограничений и нет описания экземпляра. @Vrodvi они специально были убраны при чистке ошибок?

**Reply by Vrodvi** [reply]
> Я залил файлы из input в репозиторий, дополнительно поменял title так чтобы были видны и русские, и латинские названия - это решение было принято совместно с @OlegPenzin на встрече 02.07.24. Замечена странность: в файле CoreOrganizationPayer.fsh есть только заголовки (Title, description), но нет самих ограничений и нет описания экземпляра. @Vrodvi они специально были убраны при чистке ошибок?

В input_output_для_RuCoreIG.zip файл CoreOrganizationPayer.fsh я создал пустым просто чтобы ушла ошибка со ссылкой на несуществующий ресурс, я написал об этом выше.
Полноценное наполнение для этого профиля мне не известно.

Чтобы не заниматься подобной пересылкой файлов нам надо наладить коллективную работу через репозиторий с контролем версий.
Но тогда надо дообсуждать вопрос с синхронизацией данных о профилях на сайте (обсуждали тут https://fhir.ru/meetings.2024-06-11) - хотели сделать репозиторий основным источником правды, а на сайте сделать пробрасывание этого источника (плюс решить вопрос с редактированием с сайта - в этом случае хотели делать обратное пробрасывание в репозиторий, чтобы профили хранились всегда только в одном месте).

**Reply by eugenekogan-spb** [reply]
Николай Рыжиков @niquola помог подключить [публикатор, сделанный Джошем Манделом](https://github.com/FHIR/auto-ig-builder). И вуаля: https://build.fhir.org/ig/fhir-ru/RuCoreIG/branches/main/index.html. Теперь у нас есть руководство, опубликованное на Fhir.org!
Предлагаю обсудить дальнейшие шаги на завтрашней встрече.

**Reply by Vrodvi** [reply]
Дальнейшие необходимые шаги по этой теме:

- 1) необходимо выяснить каким образом осуществлять публикацию логических версий руководств по реализации через предложеный публикатор ("релизов" руководства). Это должно быть не механическое продвижение версии, которое осущетсвляется через Git, а целенаправленное действие по публикации следующей логической версии после того как сообщество подтвердит её готовность для выпуска.
При этом всегда должна оставаться текущая рабочая версия руководства, где будут накапливаться все текущие изменения.
Поскольку экспертизы в этом вопросе у нас нет, то приветствуется консультация от коллег, сталкивавшихся с этим вопросом.

- 2) необходимо выяснить каким образом активировать в руководстве раздел с перечнем логических версий ("релизов"), который обычно открывается по ссылке "[Directory of published versions](http://hl7.org/fhir/us/core/history.html)" (тут пример для версий руководства US Core)
(начать можно с официального конфлюенса - см. https://confluence.hl7.org/pages/viewpage.action?pageId=35718627#IGPublisherDocumentation-History)
(есть некий пример шаблона для страницы с историей версий от Грэхэма - https://github.com/HL7/fhir-ig-history-template)

- 3) в реестре руководств fhir.ru - https://fhir.ru/ig/ - нужно переключить ссылки каждого из руководств на соответствующий новый репозиторий (ссылка должна вести на текущую "рабочую" версию руководства, перейдя на которую можно будет просмотреть перечень логических опубликованных версий через стандартную ссылку "Directory of published versions")

**Comment by eugenekogan-spb** [comment]
Прошу помочь разобрать ошибку, из-за которой не собирается руководство:
<img width="1261" alt="image" src="https://github.com/user-attachments/assets/19b09eba-bf84-4fbb-a28f-8f688db18c82">
Это случилось после моих изменений в fsh-текстах: изменены name в Naming System, убраны пробелы, добавлены экземпляры для двух профилей.

**Reply by eugenekogan-spb** [reply]
@Vrodvi

**Reply by eugenekogan-spb** [reply]
Можно попросить вас собрать на локальной машине текущие тексты RuCoreIG? Возможно, диагностика будет более полной.

**Reply by eugenekogan-spb** [reply]
Исправили следующим образом:

- Запустили sushi на локальной машине
- В диагностике обнаружилась ошибка - некорректный атрибут Observation.effective
- После замены на корректный Observation.effectiveDateTime руководство собралось.

На будущее нужно иметь в виду, что публикатор пишет в лог последнюю строчку из листинга sushi (ещё и обрезанную), в нашем случае это было warning по поводу русских символов в alias, в то время как найденные sushi ошибки не показаны.

**Reply by eugenekogan-spb** [reply]
написал об этом в zulip
https://chat.fhir.org/#narrow/stream/179252-IG-creation/topic/Incomplete.20auto-ig-builder.20log.20message

**Comment by eugenekogan-spb** [comment]
После очередной сборки замечено следующее некорректное поведение публикатора.

Руководство собрано автоматическим сборщиком, сообщение об успешной сборке есть на дэшборде https://fhir.github.io/auto-ig-builder/
<img width="1112" alt="image" src="https://github.com/user-attachments/assets/94fb2f2a-b104-4bc0-aabe-8ba2753793b4">

В собранном руководстве https://build.fhir.org/ig/fhir-ru/RuCoreIG/ указана корректная дата сборки - 12 августа.
<img width="678" alt="image" src="https://github.com/user-attachments/assets/4d2ecf0b-8ba4-440b-8a48-b09e87021328">


Если в руководстве нажать ссылку QA report, откроется отчет за 11 августа: https://build.fhir.org/ig/fhir-ru/RuCoreIG/qa.html
<img width="507" alt="image" src="https://github.com/user-attachments/assets/2e58630e-86d0-42ee-a37f-e7e777dfee4d">


При этом, если в строке дэшборда сборок  нажать кнопку QA, откроется правильный отчет, за 12 августа: https://build.fhir.org/ig/fhir-ru/RuCoreIG/branches/__default/qa.html
<img width="502" alt="image" src="https://github.com/user-attachments/assets/8e93b3b9-6700-4cd4-ac86-8187b407309d">

**Reply by eugenekogan-spb** [reply]
После третьей переборки само собой поправилось.
https://build.fhir.org/ig/fhir-ru/RuCoreIG/qa.html
<img width="504" alt="image" src="https://github.com/user-attachments/assets/30a9e411-8237-43b3-9908-f10fecb7fb7d">
